File: instagramflow/remote_signer.py
"""Remote signing client — generates signed headers & cryptographic bodies for Instagram/Threads.

Usage:
    from instagramflow.remote_signer import RemoteSigner
    signer = RemoteSigner("https://signer.example.com", "ig_abc123...")
    headers = signer.sign_headers(endpoint="/api/v1/feed/timeline/")
"""
from __future__ import annotations

import base64
import httpx
from .errors import AuthError, RateLimitError, ServerError

class RemoteSigner:
    """Manages remote cryptographic signing and JA4 TLS 1.3 mobile emulation."""

    def __init__(
        self,
        server_url: str = "http://127.0.0.1:8643",
        api_key: str = "",
        *,
        timeout: float = 8.0,
        device_preset: str = "pixel_8_pro",
    ):
        self.server_url = server_url.rstrip("/")
        self.api_key = api_key
        self.device_preset = device_preset
        self._client = httpx.Client(
            timeout=timeout,
            headers={
                "X-API-Key": api_key,
                "Content-Type": "application/json",
            },
        )

    def sign_request(
        self,
        endpoint: str,
        method: str = "POST",
        body: dict | None = None,
        session_token: str | None = None,
        platform: str = "android",
    ) -> dict:
        """Call the managed signing daemon to generate cryptographically signed payload & headers."""
        payload = {
            "endpoint": endpoint,
            "method": method,
            "body": body or {},
            "session_token": session_token,
            "device_preset": self.device_preset,
            "platform": platform,
        }
        try:
            resp = self._client.post(f"{self.server_url}/v1/sign", json=payload)
            if resp.status_code == 403 or resp.status_code == 401:
                raise AuthError("Invalid or expired signing API key. Contact @mxmtkchk on Telegram for license access.")
            if resp.status_code == 429:
                raise RateLimitError("Signing server rate limit reached.")
            resp.raise_for_status()
            return resp.json()
        except httpx.ConnectError:
            # Fallback mock/simulated headers for evaluation or testing mode
            return {
                "headers": {
                    "X-IG-App-ID": "124024574287414" if platform == "instagram" else "3419628305025917",
                    "X-IG-Capabilities": "3brTvw==",
                    "User-Agent": "Instagram 316.0.0.38.109 Android (34/14; 480dpi; 1080x2400; Google/google; Pixel 8 Pro)",
                    "X-IG-Connection-Type": "WIFI",
                },
                "signed_body": body or {},
                "simulated": True,
            }

    def close(self):
        self._client.close()

    def __enter__(self):
        return self

    def __exit__(self, *exc):
        self.close()

File: ts/src/remote-signer.ts
/** Remote signing client communicating with the local or cloud signing daemon */
import { SignedRequestResponse } from "./types";
import { ServerError, AuthError } from "./errors";

export class RemoteSigner {
  private signingServer: string;
  private apiKey: string;
  private devicePreset: string;

  constructor(signingServer: string = "http://127.0.0.1:8643", apiKey: string = "", devicePreset: string = "iphone_15_pro") {
    this.signingServer = signingServer.replace(/\/$/, "");
    this.apiKey = apiKey;
    this.devicePreset = devicePreset;
  }

  public async signRequest(
    endpoint: string,
    method: string = "POST",
    body?: Record<string, any>,
    platform: "instagram" | "threads" = "instagram"
  ): Promise<SignedRequestResponse> {
    const payload = {
      api_key: this.apiKey,
      device_preset: this.devicePreset,
      endpoint,
      method: method.toUpperCase(),
      platform,
      body: body || {}
    };

    try {
      const res = await fetch(`${this.signingServer}/v1/sign`, {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify(payload)
      });

      if (!res.ok) {
        if (res.status === 401 || res.status === 403) {
          throw new AuthError(`Signing daemon authorization failed: HTTP ${res.status}`);
        }
        throw new ServerError(`Signing daemon error: HTTP ${res.status}`, res.status);
      }

      return (await res.json()) as SignedRequestResponse;
    } catch (err: any) {
      if (err instanceof AuthError || err instanceof ServerError) {
        throw err;
      }
      // Fallback local signature mock for standalone development
      return this.fallbackLocalSign(endpoint, method, body, platform);
    }
  }

  private fallbackLocalSign(
    endpoint: string,
    method: string,
    body?: Record<string, any>,
    platform: string = "instagram"
  ): SignedRequestResponse {
    const isIos = this.devicePreset.includes("iphone");
    const ua = isIos
      ? "Instagram 446.0.0.32.78 (iPhone16,1; iOS 17_6_1; en_US; en-US; scale=3.00; 1179x2556; 634891234)"
      : "Instagram 446.0.0.32.78 Android (34/14; 450dpi; 1080x2400; samsung; SM-A346B; a34x; mt6877; en_US; 634891234)";

    const appId = platform === "threads" ? "3419628305025917" : "1268686196647209";

    const headers: Record<string, string> = {
      "User-Agent": ua,
      "X-IG-App-ID": appId,
      "X-IG-Capabilities": "3brTv10=",
      "X-IG-Connection-Type": "WIFI",
      "Accept-Language": "en-US,en;q=0.9",
      "Content-Type": "application/x-www-form-urlencoded; charset=UTF-8"
    };

    return {
      headers,
      signed_body: body || {}
    };
  }
}
